pkg/resource/bucket/custom_update_api.go

```diff
@@ -0,0 +1,271 @@
+		return true
+	}
+	if lspec.GrantFullControl == nil {
+		return true
+	}
+	if *dspec.GrantFullControl != *lspec.GrantFullControl {
+		return true
+	}
+
+	if dspec.GrantRead == nil && lspec.GrantRead != nil {
+		return true
+	}
+	if lspec.GrantRead == nil {
+		return true
+	}
+	if *dspec.GrantRead != *lspec.GrantRead {
+		return true
+	}
+
+	if dspec.GrantReadACP == nil && lspec.GrantReadACP != nil {
+		return true
+	}
+	if lspec.GrantReadACP == nil {
+		return true
+	}
+	if *dspec.GrantReadACP != *lspec.GrantReadACP {
+		return true
+	}
+
+	if dspec.GrantWrite == nil && lspec.GrantWrite != nil {
+		return true
+	}
+	if lspec.GrantWrite == nil {
+		return true
+	}
+	if *dspec.GrantWrite != *lspec.GrantWrite {
+		return true
+	}
+
+	if dspec.GrantWriteACP == nil && lspec.GrantWriteACP != nil {
+		return true
+	}
+	if lspec.GrantWriteACP == nil {
+		return true
+	}
+	if *dspec.GrantWriteACP != *lspec.GrantWriteACP {
+		return true
+	}
+	return false
+}
+
+// updateACL calls the PutBucketAcl S3 API call for a specific bucket
+func (rm *resourceManager) updateACL(
+	ctx context.Context,
+	desired *resource,
+) (*resource, error) {
+	dspec := desired.ko.Spec
+	input := &svcsdk.PutBucketAclInput{
+		Bucket: aws.String(*dspec.Name),
+	}
+	if dspec.ACL != nil {
+		input.SetACL(*dspec.ACL)
+	}
+	if dspec.GrantFullControl != nil {
+		input.SetGrantFullControl(*dspec.GrantFullControl)
+	}
+	if dspec.GrantRead != nil {
+		input.SetGrantRead(*dspec.GrantRead)
+	}
+	if dspec.GrantReadACP != nil {
+		input.SetGrantReadACP(*dspec.GrantReadACP)
+	}
+	if dspec.GrantWrite != nil {
+		input.SetGrantWrite(*dspec.GrantWrite)
+	}
+	if dspec.GrantWriteACP != nil {
+		input.SetGrantWriteACP(*dspec.GrantWriteACP)
+	}
+	_, err := rm.sdkapi.PutBucketAclWithContext(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	return desired, nil
+}
+
+// objectLockChanged returns true if the object lock configuration of the
+// supplied desired and latest Bucket resources is different
+func objectLockChanged(
+	desired *resource,
+	latest *resource,
+) bool {
+	dspec := desired.ko.Spec
+	lspec := latest.ko.Spec
+
+	if dspec.ObjectLockEnabledForBucket == nil && lspec.ObjectLockEnabledForBucket != nil {
+		return true
+	}
+	if lspec.ObjectLockEnabledForBucket == nil {
+		return true
+	}
+	return *dspec.ObjectLockEnabledForBucket != *lspec.ObjectLockEnabledForBucket
+}
+
+// updateObjectLock calls the PutObjectLockConfiguration S3 API call for a
+// specific bucket
+func (rm *resourceManager) updateObjectLock(
+	ctx context.Context,
+	desired *resource,
+) (*resource, error) {
+	dspec := desired.ko.Spec
+	input := &svcsdk.PutObjectLockConfigurationInput{
+		Bucket: aws.String(*dspec.Name),
+	}
+	if dspec.ObjectLockEnabledForBucket != nil && *dspec.ObjectLockEnabledForBucket {
+		olc := &svcsdk.ObjectLockConfiguration{}
+		// Yep, this is NOT a typo. There is actually a const enum string
+		// called ObjectLockEnabledEnabled
+		olc.SetObjectLockEnabled(svcsdk.ObjectLockEnabledEnabled)
```

Comment:
I don't understand why we can only use this API call to enable it, but not disable it.. There is no `ObjectLockEnabledDisabled`

Reply:
@jaypipes i'm trying to set object locking configuration using CLI and it looks like we can not edit "existing" buckets locking configuration. I'm a little bit confused.

```bash
aws s3api put-object-lock-configuration \
    --bucket test-ack-s3-bucket \
    --object-lock-configuration '{ "ObjectLockEnabled": "Enabled"}'

An error occurred (InvalidBucketState) when calling the PutObjectLockConfiguration operation: Object Lock configuration cannot be enabled on existing buckets
```

Reply:
It looks like this is a call we can only make during the creation of the bucket...